internal/gensupport/send.go

```diff
@@ -64,14 +65,51 @@ func send(ctx context.Context, client *http.Client, req *http.Request) (*http.Re
 	if client == nil {
 		client = http.DefaultClient
 	}
-	resp, err := client.Do(req.WithContext(ctx))
-	// If we got an error, and the context has been canceled,
-	// the context's error is probably more useful.
-	if err != nil {
+
+	var resp *http.Response
+	var err error
+
+	// Loop to retry the request, up to the deadline.
+	var pause time.Duration
+	bo := backoff()
+	quitAfter := time.After(retryDeadline)
+
+	for {
 		select {
 		case <-ctx.Done():
-			err = ctx.Err()
-		default:
+			// If we got an error, and the context has been canceled,
+			// the context's error is probably more useful.
+			if err == nil {
+				err = ctx.Err()
+			}
+			return resp, err
+		case <-time.After(pause):
+		case <-quitAfter:
+			return resp, err
+		}
+
+		resp, err = client.Do(req.WithContext(ctx))
+
+		var status int
+		if resp != nil {
+			status = resp.StatusCode
+		}
+
+		// Check if we can retry the request. A retry can only be done if the error
+		// is retryable and the request body can be re-created using GetBody (this
+		// will not be possible if the body was unbuffered).
+		if req.GetBody == nil || !shouldRetry(status, err) {
+			break
+		}
+		var errBody error
+		req.Body, errBody = req.GetBody()
```

Comment:
it's called on the next iteration of the for loop, assuming we don't break.

Reply:
ah, that's subtle, it's because `resp = nil` and wouldn't meet the condition at the end. Understood. This LGTM. I know you're pending more input but will give my approval.

Reply:
Ah, it doesn't actually matter if the response is nil or not-- the condition on lines 111-112 does not break out of the loop. We only break if shouldRetry returns false (which means either a good status or non-retryable error) or if we can't successfully GetBody (which generally means that the input is unbuffered and so it's impossible to make a new body reader).